p2p/service.py

```diff
@@ -112,6 +121,24 @@ def run_child_service(self, child_service: 'BaseService') -> 'asyncio.Future[Any
     def is_running(self) -> bool:
         return self._run_lock.locked()
 
+    def threadsafe_cancel(self, poll_period: float=0.1) -> None:
```

Comment:
I was surprised that this wasn't a coroutine.  Is there a reason for that?

Reply:
Hm... I could try it. It requires being cautious to make sure everything is scheduling where you intend it to: that you are `await`-ing in one loop & thread, but canceling and awaiting on a different loop and thread.